tls/s2n_async_pkey.c

```diff
@@ -641,3 +641,77 @@ S2N_RESULT s2n_async_pkey_op_copy_hash_state_for_testing(struct s2n_async_pkey_o
     RESULT_GUARD_POSIX(s2n_hash_copy(copy, &op->op.sign.digest));
     return S2N_RESULT_OK;
 }
+
+DEFINE_POINTER_CLEANUP_FUNC(struct s2n_hash_state *, s2n_hash_free);
+
+DEFINE_POINTER_CLEANUP_FUNC(struct s2n_blob *, s2n_free);
+
+static S2N_RESULT s2n_async_pkey_verify_perform(struct s2n_async_op *op)
+{
+    RESULT_ENSURE_REF(op);
+    RESULT_ENSURE_REF(op->conn);
+    RESULT_ENSURE(op->type == S2N_ASYNC_PKEY_VERIFY, S2N_ERR_INVALID_ARGUMENT);
+
+    struct s2n_pkey *pub_key = NULL;
+    if (op->conn->mode == S2N_CLIENT) {
+        pub_key = &op->conn->handshake_params.server_public_key;
+    } else {
+        pub_key = &op->conn->handshake_params.client_public_key;
+    }
+
+    struct s2n_async_pkey_verify_data *verify = &op->op_data.async_pkey_verify;
+    /* Free the memory allocated in s2n_async_pkey_verify_async() */
+    DEFER_CLEANUP(struct s2n_hash_state *digest = &verify->digest, s2n_hash_free_pointer);
+    DEFER_CLEANUP(struct s2n_blob *signature = &verify->signature, s2n_free_pointer);
```

Comment:
The memory management here seems a little odd, I think because this method doesn't actually declare digest/signature OR allocate digest/signature.

I wonder if we need a dedicated free/wipe method for each op type? Like s2n_async_pkey_verify_data_wipe? Or even one shared s2n_async_op/s2n_async_offload_op wipe method that branches on type? For the former the free would still happen in this method, since this is where we know the op type. But for the latter the free could happen in s2n_async_op_reset and on connection wipe. If we can also handle the free on connection wipe, then we might not have to be as strict about freeing on error, because eventually s2n_connection_wipe will clean up the leftovers.